File: 01/cpp/second.cpp
/*
 * Compile using gcc
 * 
 * $ g++ ./second.cpp -o second.out
 * 
 * Run directly
 * 
 * $ cat input.txt | time ./cpp/second.out
 */

#include <iostream>
#include <string>
#include <string_view>
#include <utility>

constexpr std::pair<std::string_view, int> mappings[19] = {
    {"0", 0},
    {"1", 1},
    {"2", 2},
    {"3", 3},
    {"4", 4},
    {"5", 5},
    {"6", 6},
    {"7", 7},
    {"8", 8},
    {"9", 9},
    {"one", 1},
    {"two", 2},
    {"three", 3},
    {"four", 4},
    {"five", 5},
    {"six", 6},
    {"seven", 7},
    {"eight", 8},
    {"nine", 9},
};

int main() {
    std::string s;
    int n;
    while (true) {
        std::getline(std::cin, s);
        std::string_view sv(s);
        for (int i = 0; i < sv.size(); i++) {
            std::string_view subsv = sv.substr(i);
            for (int j = 0; j < sizeof mappings / sizeof mappings[0]; j++) {
                auto [matchstr, num] = mappings[j];
                if (matchstr.size() > subsv.size()) continue;
                std::string_view matchsubsv = subsv.substr(0, matchstr.size());
                if (matchsubsv == matchstr) {
                    n += num * 10;
                    goto second;
                }
            }
        }
        break;
    second:
        for (int i = s.size(); i > 0; i--) {
            std::string_view subsv = sv.substr(i - 1);
            for (int j = 0; j < sizeof mappings / sizeof mappings[0]; j++) {
                auto [matchstr, num] = mappings[j];
                if (matchstr.size() > subsv.size()) continue;
                std::string_view matchsubsv = subsv.substr(0, matchstr.size());
                if (matchsubsv == matchstr) {
                    n += num;
                    goto next;
                }
            }
        }
        break;
    next:
        s.clear();
    }
    std::cout << n << std::endl;
}

File: 01/js/second.js
/*
 * Run using deno
 *
 * $ cat input.txt | time deno run ./js/second.js
 */

import { readLines } from "https://deno.land/std/io/mod.ts";
let t=0,z
let replacements = {
	"one": 1,
	"two": 2,
	"three": 3,
	"four": 4,
	"five": 5,
	"six": 6,
	"seven": 7,
	"eight": 8,
	"nine": 9,
};
for await (const l of readLines(Deno.stdin)) {
	if (!l) break;
	let s = l;
	for (let i = 0; i < s.length; i++) {
		let start = s.substr(0, i);
		let rest = s.substr(i);
		for (const [key, value] of Object.entries(replacements)) {
			if (rest.indexOf(key) == 0) {
				rest=rest.replace(key, value);
				i=0;
			}
		}
		s = `${start}${rest}`;
	}
	z=s.split``.filter(n=>!Number.isNaN(+n)),t+=+`${z[0]}${z[z.length-1]}`
}
console.log(t)
